provider/pkg/schema/gen.go

```diff
@@ -933,45 +933,35 @@ func (ctx *context) propertyTypeSpec(parentName string, propSchema *jsschema.Sch
 		return &pschema.TypeSpec{Ref: referencedTypeName}, nil
 	}
 
-	// Union types.
-	if len(propSchema.AnyOf) == 1 {
-		return ctx.propertyTypeSpec(parentName, propSchema.AnyOf[0])
-	}
-
-	if len(propSchema.AnyOf) > 1 {
-		var types []pschema.TypeSpec
-		for _, sch := range propSchema.AnyOf {
-			typ, err := ctx.propertyTypeSpec(parentName, sch)
+	// Union types
+	if len(propSchema.AnyOf) > 0 || len(propSchema.OneOf) > 0 {
+		schemas := FlattenJSSchema(propSchema)
+		types := make([]pschema.TypeSpec, 0, len(schemas))
+		typeMap := make(map[string]bool, len(schemas))
+		for i, sch := range schemas {
+			typ, err := ctx.propertyTypeSpec(fmt.Sprintf("%s%d", parentName, i), sch)
 			if err != nil {
 				return nil, err
 			}
-			types = append(types, *typ)
-		}
-		return &pschema.TypeSpec{
-			OneOf: types,
-		}, nil
-	}
-
-	if len(propSchema.OneOf) == 1 {
-		return ctx.propertyTypeSpec(parentName, propSchema.OneOf[0])
-	}
 
-	if len(propSchema.OneOf) > 1 {
-		var types []pschema.TypeSpec
-		for i, sch := range propSchema.OneOf {
-			typ, err := ctx.propertyTypeSpec(fmt.Sprintf("%s%d", parentName, i), sch)
+			// stringify to create a key for identifying duplicates
+			keybytes, err := json.Marshal(*typ)
```

Comment:
I don't believe that unions have ordered keys, so this could allow duplicates to slip through. I will let @danielrbradley decide if that matters.